experimental/Pipeline/pipelie.mli

```diff
@@ -0,0 +1,204 @@
+
+(********************************************************************************)
+(*** Modules that belong to different files *)
+(********************************************************************************)
+module Feedback : sig
+  type t
+end
+
+module Trace : sig
+  type t
+end
+
+module Context : sig
+  type 'a t
+end
+
+module Sexp : sig
+  type t
+end
+
+module NameSymbol : sig
+  type t
+end
+
+(********************************************************************************)
+(** Meta Information *)
+(********************************************************************************)
+
+module Meta : sig
+  type t = { feedback : Feedback.t; trace : Trace.t }
+end
+
+
+(********************************************************************************)
+(** Step Information Information *)
+(********************************************************************************)
+
+module Step : sig
+  type 'a t =
+    | Success of {meta : Meta.t; result           : 'a}
+    | Failure of {meta : Meta.t; partial_feedback : 'a option}
+end
+
+(********************************************************************************)
+(** Pipeline *)
+(********************************************************************************)
+
+(** A Recursive list represents the concept of a list with a chance to nest
+ * This allows us to get a tree like structure,
+ * where A pipeline step can contain more steps
+ *)
+module RecursiveList : sig
+  type 'a t
+end
+
+(** A circular list represents the idea of steps that can be recursive
+ * on each other *)
+module CircularList : sig
+  (* TODO : add a recursion function to the recursive bit, that tells
+   * it how it's recursive. perhaps it uses *)
+  type 'a recursive_schema =
+    | Recursive of 'a list
+    | NonRecursive of 'a
+
+  type 'a t =
+    'a recursive_schema RecursiveList.t
+end
+
+module Pipeline : sig
+  (** [t] is the Pipeline argument itself. This is what is actively
+   * being processed in the pipeline itself.
+   *)
+  type t = {
+      current_expression : Sexp.t list;
+      context            : Sexp.t Context.t
+    }
+
+  type has_reduce = HasReduce
+
+  (* HasReduce m =>  *)
```

Comment:
Since OCaml doesn't have effects, I have to put the type class annotation in comments for now. Though I still have design around this as I want to decide what to do with the `eff` here, as It's unclear if It'd work in Haskell to have a function that can satisfy this constraint... something I'll have to think of and rewrite if not!